src/librustc/sir.rs

```diff
@@ -1,174 +1,112 @@
-            res[func_idx] = IndexVec::from_elem_n(None, func.blocks.len());
-        }
-
-        // Now we iterate over our hash table, replacing the aforementioned `None`s.
-        for (bb, (func_idx, bb_idx)) in &self.llvm_blocks {
-            debug_assert!(res[*func_idx][*bb_idx] == None);
-            res[*func_idx][*bb_idx] = Some(*bb);
-        }
-
-        // Now get rid of the Option wrappers around the pointers. The `unwrap()` is guaranteed to
-        // succeed, as the above loop mutates every single `None` to a `Some`.
-        let mut ret = IndexVec::default();
-        for func_blocks in res.into_iter() {
-            ret.push(func_blocks.into_iter().map(|b| b.unwrap()).collect());
-        }
-
-        ret
-    }
+/// A structure for building the SIR of a function.
+pub struct SirFuncCx {
+    pub func: ykpack::Body,
+}
 
-    /// Dump SIR to text file.
-    /// Used in tests and for debugging.
-    pub fn dump(&self, dest: &mut dyn Write) -> Result<(), io::Error> {
-        for func in &self.funcs {
-            writeln!(dest, "{}", func)?;
+impl SirFuncCx {
+    pub fn new<'tcx>(tcx: TyCtxt<'tcx>, instance: &Instance<'tcx>, num_blocks: usize) -> Self {
+        let symbol_name = format!("new__{}", tcx.symbol_name(*instance).name.as_str());
+
+        let mut flags = 0;
+        for attr in tcx.get_attrs(instance.def_id()).iter() {
+            if attr.check_name(sym::trace_head) {
+                flags |= ykpack::bodyflags::TRACE_HEAD;
+            } else if attr.check_name(sym::trace_tail) {
+                flags |= ykpack::bodyflags::TRACE_TAIL;
+            }
         }
 
-        Ok(())
+        // Since there's a one-to-one mapping between MIR and SIR blocks, we know how many SIR
+        // blocks we will need and can allocate empty SIR blocks ahead of time.
+        let blocks = vec![
+            ykpack::BasicBlock {
+                stmts: Default::default(),
+                term: ykpack::Terminator::Unreachable,
+            };
+            num_blocks
+        ];
+
+        Self { func: ykpack::Body { symbol_name, blocks, flags } }
     }
 
-    /// Given an llvm::BasicBlock returns the equivalent ykpack::BasicBlock in the SIR.
-    fn get_sir_block(&mut self, bb: *const BasicBlock) -> &mut ykpack::BasicBlock {
-        let (sirfuncidx, sirblockidx) = self.llvm_blocks[&bb];
-        let sir_func = &mut self.funcs[sirfuncidx];
-        &mut sir_func.blocks[sirblockidx.index()]
+    /// Returns true if there are no basic blocks.
+    pub fn is_empty(&self) -> bool {
+        self.func.blocks.len() == 0
     }
 
-    /// Set the current position of builder to `pos`. Equivalent to LLVMPositionBuilderBefore.
-    pub fn position_before(&mut self, builder: *const Builder, bb: *const BasicBlock, pos: usize) {
-        self.builders.insert(builder, (bb, pos));
+    /// Appends a statement to the specified basic block.
+    fn append_stmt(&mut self, bb: ykpack::BasicBlockIndex, stmt: ykpack::Statement) {
```

Comment:
`push_stmt` would be a bit more Rusty-y a name?

Reply:
Agreed.